### PR TITLE
feat: integrate Advanced search criteria with New Search UI

### PR DESCRIPTION
**Pull request by PenghaiZhang**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [ ] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This is the fourth (and hopefully the last) PR for the advanced search criteria. It's all about the integration with `SearchPage`.
 
Main changes are:
1. Add a new field to `SearchParam` and `SearchOptions` to support the raw Lucene query.
2. Hide the Adv search panel whenever a search is triggered.
3. Update how to generate the initial advanced search query values.
4. Rework how to perform the initial search.
5. Update how to check if any adv search criteria has been set.

### CODE REVIEW

File: react-front-end/__tests__/tsrc/search/AdvancedSearch.test.tsx [hunk 1]
```diff
@@ -293,3 +298,26 @@ describe("Rendering of wizard", () => {
...
+describe("search with Advanced search criteria", () => {
+  it("supports searching with raw Lucene query", async () => {
+    jest.clearAllMocks();
+    const defaultValue = "hello world";
+    mockGetAdvancedSearchByUuid.mockResolvedValue(
+      oneEditBoxWizard(false, [defaultValue])
+    );
+    const { container } = await renderAdvancedSearchPage();
+
+    // Search with the EditBox's default value.
```
Comment (edalex-ian) [inline react-front-end/__tests__/tsrc/search/AdvancedSearch.test.tsx:311]: Hmmm, but doesn't that now happen on initial render?

File: react-front-end/tsrc/modules/SearchModule.ts [hunk 2]
```diff
@@ -126,6 +127,14 @@ export interface SearchOptions {
...
    */
   musts?: OEQ.Search.Must[];
+  /**
+   * Raw Lucene query generated from the values of Wizard controls.
+   */
+  customLuceneQuery?: string;
+  /**
+   * Currently configured Advanced search criteria.
+   */
+  advFieldValue?: FieldValueMap;
```
Comment (edalex-ian) [inline react-front-end/tsrc/modules/SearchModule.ts:137]: Are these `SearchOptions` or `SearchPageOptions`?

File: react-front-end/tsrc/search/SearchPage.tsx [hunk 3]
```diff
@@ -295,33 +305,53 @@ const SearchPage = ({ updateTemplate, advancedSearchId }: SearchPageProps) => {
...
+            return initialQueryValues;
+          };
+
+          const initialRawLuceneQuery: string | undefined = await pipe(
+            O.fromNullable(advancedSearchDefinition),
+            O.map(
+              flow(getInitialAdvancedSearchQueryValues, generateRawLuceneQuery)
+            ),
+            O.getOrElseW(() => Promise.resolve(undefined))
+          );
```
Comment (edalex-ian) [inline react-front-end/tsrc/search/SearchPage.tsx:332]: In a continued bid to keep this effect and SearchPage.tsx under control, this should probably be placed in a function in SearchPageHelper.tsx. And then would it be possible to place that function in the array of functions for the `Promise.all` block? Or (acknowledging dependencies on the output from other items in that array), how about simply changing the line in setting `customeLuceneQuery` of `initialSearchPageOptions` to call the new function (with `await`)?

File: react-front-end/tsrc/search/SearchPage.tsx [hunk 4]
```diff
@@ -738,8 +799,13 @@ const SearchPage = ({ updateTemplate, advancedSearchId }: SearchPageProps) => {
...
   const isAdvSearchCriteriaSet = (queryValues: FieldValueMap): boolean => {
-    const isAnyFieldSet: boolean = Array.from(queryValues).some(
-      ([_, values]) => values.length > 0
+    const isAnyFieldSet = pipe(
+      queryValues,
+      // Some controls like Calendar may have an empty string as their default values which should be
+      // filtered out.
+      M.map(A.filter(isNonEmptyString)),
+      M.values<ControlValue>(OrdAsIs),
+      A.some((v) => v.length > 0)
```
Comment (edalex-ian) [inline react-front-end/tsrc/search/SearchPage.tsx:808]: Function should sit over in `WizardHelper` me thinks. :wink:

**Review by edalex-ian — CHANGES_REQUESTED** [review]
Lookin' good @PenghaiZhang . A couple of queries, and a couple of requests for shuffling. :wink:

File: react-front-end/__tests__/tsrc/search/AdvancedSearch.test.tsx (hunk 1 above)
Comment (PenghaiZhang) [reply react-front-end/__tests__/tsrc/search/AdvancedSearch.test.tsx:311]: Ah yeah! Do not need to click the button at all.

File: react-front-end/tsrc/search/SearchPage.tsx (hunk 4 above)
Comment (PenghaiZhang) [reply react-front-end/tsrc/search/SearchPage.tsx:808]: Ah yes! `isControlValueNonEmpty` I found it.

File: react-front-end/tsrc/modules/SearchModule.ts (hunk 2 above)
Comment (PenghaiZhang) [reply react-front-end/tsrc/modules/SearchModule.ts]: `advFieldValue` has been moved `SearchPageOptions`.

File: react-front-end/tsrc/search/SearchPage.tsx (hunk 4 above)
Comment (edalex-ian) [reply react-front-end/tsrc/search/SearchPage.tsx]: As discussed, once merged then this can be refactored in another PR. :wink:

**Review by edalex-ian — APPROVED** [review]
As discussed, all looks good and I'm happy to wait for that last bit to be refactored as part of another PR.

**Comment by edalex-ian** [comment]
Will merge now to keep the ball rolling so that @edalex-yinzi can merge to his clear button branch to finish that feature.

**Pull request merged**